include/seastar/core/loop.hh

```diff
@@ -563,8 +563,9 @@ parallel_for_each(Iterator begin, Sentinel end, Func&& func) noexcept {
     //   - available, failed: collect exception in ex
     //   - not available: collect in s (allocating it if needed)
     while (begin != end) {
-        auto f = futurize_invoke(std::forward<Func>(func), *begin);
+        auto tmp = begin;
         ++begin;
+        auto f = futurize_invoke(std::forward<Func>(func), *tmp);
```

Comment:
I don't understand how this prevents iterator invalidation? We require the iterators to be valid for the duration of the `parallel_for_each()`. Avoiding it being invalid here, will not prevent dereferencing below.

Reply:
According to this commit https://github.com/scylladb/seastar/commit/b1b0bff1b8b3a94792f7a37b203ae749d54a1d83, It says "applying postfix++ on a move-only iterator returns void, so *it++ does not work;". Maybe *begin++ isn't an option. What's your suggestion?

Reply:
@rosinL could you point me to the caller site of `parallel_for_each()` where you spotted the segment fault? mutating the reference of an object returned by dereferencing an iterator should not invalidate the iterator, if that's the case, that'd be a bad smell in the code. and probably, we need to revisit the caller.

in my opinion, to make a **copy** of a move-only iterator would defeat the purpose of b1b0bff1b8b3a94792f7a37b203ae749d54a1d83 which wants to ensure that `parallel_for_each()` does is able to work with move-only ranges represented by move-only iterators.

Reply:
> in my opinion, to make a **copy** of a move-only iterator would defeat the purpose of [b1b0bff](https://github.com/scylladb/seastar/commit/b1b0bff1b8b3a94792f7a37b203ae749d54a1d83) which wants to ensure that `parallel_for_each()` does is able to work with move-only ranges represented by move-only iterators.

for the very reason, this change fails the build of `coroutines_test.cc`, see

https://app.circleci.com/pipelines/github/scylladb/seastar/1338/workflows/43f5880d-b6d8-4399-af1e-1e07967b3ebf/jobs/2961?invite=true#step-105-488

Reply:
In func `seastar::future<> SocketMessenger::shutdown()` use `parallel_for_each` to close and delete connects. @tchaikov The segment fault can be reproduction with perf-crimson-msgr(in shutdown phase, release version has higher probability) . I have repalce  `parallel_for_each` with `do_for_each`, it also works well, but it may not be an good choice to change to `do_for_each`.

Reply:
@rosinL hi Rixin, will take a look at the test failure. IIRC, you’ve been working on aarch64 arch. can the test failure be reproduced on amd64 machines?

Reply:
I have tested on x86 and could be reproduced. I add a ceph tracker, link is https://tracker.ceph.com/issues/56589 @tchaikov

Reply:
Perhaps we need a separate overload for the two use cases.

Reply:
@avikivity i'd suggest close this PR. it is a bug in crimson. the way how we use `parallel_for_each()` is something like
```c++
class Messenger {
  std::set<ConnPtr> connections;
public:
  future<> shutdown_all() {
    return parallel_for_each(this->connections, [](auto conn) {
      proto->close(conn);
    });
  }
};

class Proto {
  void close(ConnPtr& conn) {
    messenger->connections.erase(conn);
  }
};
```

in other words; it's like
```c++
std::set<seastar::shared_ptr<Conn>> conns;
for (auto& conn : conns) {
  conns.erase(conn);
}
```

this is a no-no in STL. i think it's also an unsupported use case in Seastar.

Reply:
FWIW, should be fixed by https://github.com/ceph/ceph/pull/47152

Reply:
> Perhaps we need a separate overload for the two use cases.
I agree, we need two kinds of parallel_for_each for  move-only iterator and copyable iterator.

Reply:
discussed with @rosinL offline. i still think we should not bend Seastar to support to this use case.

IMHO, it'd be difficult to formalize this requirement in the function signature either using concept or runtime assertion.

Reply:
I wonder if we can have a new primitive (coroutine-based) to support fork/join.

```c++
future<> do_something_to_vector(std::vector<int> v) {
     coroutine::join j;
     j.reserve(v.size()); // optional
     for (auto i : v) {
          if (std::is_odd(i)) {
              j.add([] (int i) -> future<> { ... };
          }
     }
     co_await j;
}
```

So instead of parallel_for_each controlling both iteration and joining, iteration is under the calling coroutine control, and joining is performed by coroutine::join.

Reply:
> discussed with @rosinL offline. i still think we should not bend Seastar to support to this use case.

Then please close the pull.